alfred-telemetry-platform/src/main/java/eu/xenit/alfred/telemetry/binder/solr/sharding/SolrShardingMetrics.java

```diff
@@ -47,19 +48,37 @@ public void updateMetrics() {
             solrShardingMetricsContainer.getShards(floc).forEach(shard -> {
                 Tags shardTags = flocTags.and(Tags.of("shard", String.valueOf(shard.getInstance())));
                 Set<ShardInstance> shardInstances = solrShardingMetricsContainer.getShardInstances(shard);
-                int numberOfActiveShardInstances = (int) shardInstances.stream().filter(shardInstance ->
-                        solrShardingMetricsContainer.getReplicaState(shardInstance) == ReplicaState.ACTIVE).count();
+                int numberOfActiveShardInstances = (int) shardInstances.stream()
+                        .filter(shardInstance -> solrShardingMetricsContainer.getReplicaState(shardInstance).isPresent())
+                        .filter(shardInstance -> solrShardingMetricsContainer.getReplicaState(shardInstance).get() == ReplicaState.ACTIVE)
+                        .count();
                 setAndCreateMetricIfNotExists("activeShardInstances", numberOfActiveShardInstances,
                         shardTags, "number");
                 shardInstances.forEach(shardInstance -> {
                     Tags instanceTags = shardTags.and(Tags.of("instanceHost", shardInstance.getHostName()));
-                    ShardState shardState = solrShardingMetricsContainer.getShardState(shardInstance);
+                    ShardState shardState = solrShardingMetricsContainer.getShardState(shardInstance).orElseGet(() -> {
+                        ShardState state = new ShardState();
+                        state = new ShardState();
+                        state.setLastIndexedChangeSetId(-1);
+                        state.setLastIndexedTxId(-1);
+                        state.setLastIndexedChangeSetCommitTime(-1);
+                        state.setLastIndexedTxCommitTime(-1);
+                        state.setMaster(false);
+                        state.setLastUpdated(-1);
+                        return state;
+                    });
                     setAndCreateMetricIfNotExists("lastIndexedChangeSetId",
                             shardState.getLastIndexedChangeSetId(), instanceTags, "number");
                     setAndCreateMetricIfNotExists("lastIndexedTxId", shardState.getLastIndexedTxId(),
                             instanceTags, "number");
-                    setAndCreateMetricIfNotExists("instanceMode",
-                            solrShardingMetricsContainer.getReplicaState(shardInstance).ordinal(), instanceTags, "enumValue");
+                    Optional<ReplicaState> replicaStateOpt = solrShardingMetricsContainer
+                            .getReplicaState(shardInstance);
+                    //instance mode 0 means active, 1 means lagging and 2 means silent.
+                    int instanceMode = -1;
+                    if (!replicaStateOpt.isEmpty()) {
+                        instanceMode = solrShardingMetricsContainer.getReplicaState(shardInstance).get().ordinal();
```

Comment:
I did not see this before, but you already have the ReplicaState in an Optional above. Why are you fetching it again here?

Reply:
Well cause we have to count() how many shardstates are active in the entire shardinstances list and this isn't that easy to do inside the next forEach.
It's a trade-off I guess. I'll make a counter variable and use an AtomicInteger so that we don't have to fetch it again. ;)